nmigen_boards/mercury.py

```diff
@@ -0,0 +1,226 @@
+            Attrs(IOSTANDARD="LVTTL")
+        ),
+
+        # FPGA is primary.
+        *SPIFlashResources(0,
+            cs="P27", clk="P53", mosi="P46", miso="P51",
+            attrs=Attrs(IO_STANDARD="LVTTL")
+        ),
+
+        # ADC over SPI- FPGA is primary.
+        Resource("adc", 0,
+            Subsignal("cs", PinsN("P12", dir="o")),
+            Subsignal("clk", Pins("P9", dir="o")),
+            Subsignal("mosi", Pins("P10", dir="o")),
+            Subsignal("miso", Pins("P21", dir="i")),
+            Attrs(IOSTANDARD="LVTTL")
+        ),
+
+        # GPIO control- SRAM and connectors are shared: these pins control how
+        # to access each. Recommended to combine with gpio_sram_bus extension,
+        # since these pins are related but not exposed on connectors.
+        Resource("gpio_ctl", 0,
+            Subsignal("ce", PinsN("P3", dir="o")),  # Memory chip-enable.
+            # Called MEM_CEN in schematic.
+            Subsignal("bussw_oe", Pins("P30N", dir="o")),  # 5V tolerant GPIO
+            # is shared w/ memory using this pin.
+            Attrs(IOSTANDARD="LVTTL", SLEW="FAST")
+        )
+    ]
+
+    # Perhaps define some connectors as having a specific purpose- i.e. a 5V GPIO
+    # bus with data, peripheral-select, and control signals?
+    connectors = [
+        Connector("gpio", 0, """P59 P60 P61 P62 P64 P57
+                        P56 P52 P50 P49 P85 P84
+                        P83 P78 P77 P65 P70 P71
+                        P72 P73 P5 P4 P6 P98
+                        P94 P93 P90 P89 P88 P86"""),  # 5V I/O- LVTTL.
+        Connector("dio", 0, "P20 P32 P33 P34 P35 P36 P37"),  # Fast 3.3V IO
+        # (Directly attached to FPGA)- LVCMOS33.
+        Connector("clkio", 0, "P40 P44"),  # Clock IO (Can be used as GPIO)-
+        # LVCMOS33.
+        Connector("input", 0, "P68 P97 P7 P82"),  # Input-only pins- LVCMOS33.
+        Connector("led", 0, "P13 P15 P16 P19"),  # LEDs can be used as pins
+        # as well- LVTTL.
+        Connector("pmod", 0, "P5 P4 P6 P98 P94 P93 P90 P89")  # Baseboard PMOD.
+        # Overlaps w/ GPIO bus.
+    ]
+
+    # Some default useful extensions. Attach to platform using:
+    # p.add_resources(p.leds)
+    # pmod_btn = plat.request("user_led")
+    leds = [
+        Resource("user_led", 0, Pins("1", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_led", 1, Pins("2", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_led", 2, Pins("3", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_led", 3, Pins("4", dir="o", conn=("led", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+    ]
+
+    # SRAM and 5V-tolerant I/O share a parallel bus on 200k gate version.
+    # The SRAM controller needs to take care of switching the bus between
+    # the two. Meant to be Cat() into one GPIO bus, and combined with gpio_ctl.
```

Comment:
This is... really, really confusing. Just reading this comment I have absolutely no idea how to use it correctly. Perhaps this needs some rethinking?

Reply:
Put another way:

The  `gpio_ctl` `Resource` shouldn't exist at all. It is only useful when paired with the `gpio_sram_bus` `Resource`. 

`gpio_sram_bus` and essentially the whole `gpio` connector are electrically connected, so using the SRAM implies giving up the `gpio` connector. 

The remaining two pins provided by `gpio_ctl` are _not_ broken out to any connector. `omigen` couldn't handle a peripheral that was only-partially broken-out to a connector. So I added that comment to let ppl know (apparently badly) that "you really can't use `gpio_sram_bus` without `gpio_ctl`.

Can `nmigen` handle a only-partially broken-out peripheral as a single `Resource`?

Reply:
>Can nmigen handle a only-partially broken-out peripheral as a single `Resource`?

The following `Resource` definition appears to "do the correct thing":

```
    gpio_sram_bus = [
        Resource("gpio_sram_bus", 0,
            Subsignal("ce", PinsN("P3", dir="o")),  # Memory chip-enable.
            # Called MEM_CEN in schematic.
            Subsignal("bussw_oe", Pins("P30N", dir="o")),  # 5V tolerant GPIO
            # is shared w/ memory using this pin.

            # Actual SRAM connections.
            Subsignal("a", Pins(""" 1  2  3  4  5  6  7  8  9 10
                                   11 12 13 14 15 16 17 18 19 20""", dir="o",
                                conn=("gpio", 0))),
            # A19 is actually unused- free for GPIO
            # 8-bit data bus
            Subsignal("d", Pins("21 22 23 24 25 26 27 28", dir="io",
                                conn=("gpio", 0))),
            Subsignal("we", PinsN("29", dir="o", conn=("gpio", 0))),
            # "Unused" is only used by GPIO, but is included here for
            # consistency.
            Subsignal("unused", Pins("30", dir="io", conn=("gpio", 0))),
            # Subsignal("oe_n", Pins()),  # If OE wasn't tied to ground on
            # Mercury, this pin would be here.
            Attrs(IOSTANDARD="LVTTL", SLEW="FAST")
        )
    ]
```

From playing in a REPL:

```
William@William-THINK MINGW64 ~/src/migen_mercury/baseboard
$ python3
Python 3.7.3 (default, Apr 23 2019, 08:20:20)  [GCC 8.3.0 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from nmigen_boards.mercury import *
>>> plat = MercuryPlatform()
>>> plat.add_resources(plat.gpio_sram_bus)
>>> sigs = plat.request("gpio_sram_bus")
>>> sigs
(rec gpio_sram_bus_0 (rec gpio_sram_bus_0__ce o) (rec gpio_sram_bus_0__bussw_oe o) (rec gpio_sram_bus_0__a o) (rec gpio_sram_bus_0__d i o oe) (rec gpio_sram_bus_0__we o) (rec gpio_sram_bus_0__unused i o oe))
```

From what I can gather, this works because in `nmigen`, adding a `Resource` is done the same way (`plat.add_resource`) regardless of whether said `Resource` uses connector pins or not. So it makes sense that mixing and matching `Pins` that resolve to connectors and `Pins` that resolve directly to FPGA pins works.

I wonder if this was _also_ true in `omigen` and I never noticed it?